README.md
# DDNS_Generic Library

[![arduino-library-badge](https://www.ardu-badge.com/badge/DDNS_Generic.svg?)](https://www.ardu-badge.com/DDNS_Generic)
[![GitHub release](https://img.shields.io/github/release/khoih-prog/DDNS_Generic.svg)](https://github.com/khoih-prog/DDNS_Generic/releases)
[![GitHub](https://img.shields.io/github/license/mashape/apistatus.svg)](https://github.com/khoih-prog/DDNS_Generic/blob/master/LICENSE)
[![contributions welcome](https://img.shields.io/badge/contributions-welcome-brightgreen.svg?style=flat)](#Contributing)
[![GitHub issues](https://img.shields.io/github/issues/khoih-prog/DDNS_Generic.svg)](http://github.com/khoih-prog/DDNS_Generic/issues)

<a href="https://www.buymeacoffee.com/khoihprog6" title="Donate to my libraries using BuyMeACoffee"><img src="https://cdn.buymeacoffee.com/buttons/v2/default-yellow.png" alt="Donate to my libraries using BuyMeACoffee" style="height: 50px !important;width: 181px !important;" ></a>
<a href="https://www.buymeacoffee.com/khoihprog6" title="Donate to my libraries using BuyMeACoffee"><img src="https://img.shields.io/badge/buy%20me%20a%20coffee-donate-orange.svg?logo=buy-me-a-coffee&logoColor=FFDD00" style="height: 20px !important;width: 200px !important;" ></a>


---
---

## Table of Contents

* [Why do we need this DDNS_Generic library](#why-do-we-need-this-ddns_generic-library)
  * [Features](#features)
  * [Currently Supported Boards](#currently-supported-boards)
  * [Currently Supported WiFi Modules/Shields](#currently-supported-wifi-modulesshields)
  * [Currently Supported Ethernet Modules/Shields](#currently-supported-ethernet-modulesshields)
* [Changelog](changelog.md)
* [Prerequisites](#prerequisites)
* [Installation](#installation)
  * [Use Arduino Library Manager](#use-arduino-library-manager)
  * [Manual Install](#manual-install)
  * [VS Code & PlatformIO](#vs-code--platformio)
* [Packages' Patches](#packages-patches)
  * [1. For Adafruit nRF52840 and nRF52832 boards](#1-for-adafruit-nRF52840-and-nRF52832-boards)
  * [2. For Teensy boards](#2-for-teensy-boards)
  * [3. For Arduino SAM DUE boards](#3-for-arduino-sam-due-boards)
  * [4. For Arduino SAMD boards](#4-for-arduino-samd-boards)
      * [For core version v1.8.10+](#for-core-version-v1810)
      * [For core version v1.8.9-](#for-core-version-v189-)
  * [5. For Adafruit SAMD boards](#5-for-adafruit-samd-boards)
  * [6. For Seeeduino SAMD boards](#6-for-seeeduino-samd-boards)
  * [7. For STM32 boards](#7-for-stm32-boards) 
    * [7.1. For STM32 boards to use LAN8720](#71-for-stm32-boards-to-use-lan8720)
    * [7.2. For STM32 boards to use Serial1](#72-for-stm32-boards-to-use-serial1)
  * [8. For RP2040-based boards using Earle Philhower arduino-pico core](#8-for-rp2040-based-boards-using-earle-philhower-arduino-pico-core)
    * [8.1. To use BOARD_NAME](#81-to-use-board_name)
    * [8.2. To avoid compile error relating to microsecondsToClockCycles](#82-to-avoid-compile-error-relating-to-microsecondstoclockcycles)
  * [9. For Portenta_H7 boards using Arduino IDE in Linux](#9-for-portenta_h7-boards-using-arduino-ide-in-linux)
  * [10. For RTL8720DN boards using AmebaD core](#10-for-rtl8720dn-boards-using-amebad-core)
  * [11. For SAMD21 and SAMD51 boards using ArduinoCore-fab-sam core](#11-For-SAMD21-and-SAMD51-boards-using-ArduinoCore-fab-sam-core)
  * [12. For Seeeduino RP2040 boards](#12-For-Seeeduino-RP2040-boards)
* [Libraries' Patches](#libraries-patches)
  * [1. For application requiring 2K+ HTML page](#1-for-application-requiring-2k-html-page)
  * [2. For Ethernet library](#2-for-ethernet-library)
  * [3. For EthernetLarge library](#3-for-ethernetlarge-library)
  * [4. For Etherne2 library](#4-for-ethernet2-library)
  * [5. For Ethernet3 library](#5-for-ethernet3-library)
  * [6. For UIPEthernet library](#6-for-uipethernet-library)
  * [7. For fixing ESP32 compile error](#7-for-fixing-esp32-compile-error)
  * [8. For STM32 core F3 and F4 using UIPEthernet library](#8-for-stm32-core-f3-and-f4-using-uipethernet-library)
* [HOWTO Use analogRead() with ESP32 running WiFi and/or BlueTooth (BT/BLE)](#howto-use-analogread-with-esp32-running-wifi-andor-bluetooth-btble)
  * [1. ESP32 has 2 ADCs, named ADC1 and ADC2](#1--esp32-has-2-adcs-named-adc1-and-adc2)
  * [2. ESP32 ADCs functions](#2-esp32-adcs-functions)
  * [3. ESP32 WiFi uses ADC2 for WiFi functions](#3-esp32-wifi-uses-adc2-for-wifi-functions)
* [Configuration Notes](#configuration-notes)
  * [1. How to use built-in WiFi in ESP8266/ESP32](#1-how-to-use-built-in-wifi-in-esp8266esp32)
  * [2. How to select which built-in Ethernet or shield to use](#2-how-to-select-which-built-in-ethernet-or-shield-to-use)
  * [3. How to select which built-in WiFi or shield to use](#3-how-to-select-which-built-in-wifi-or-shield-to-use)
  * [Important](#important)
  * [4. How to select another CS/SS pin to use](#4-how-to-select-another-csss-pin-to-use)
  * [5. How to use W5x00 with ESP8266](#5-how-to-use-w5x00-with-esp8266)
  * [6. How to increase W5x00 TX/RX buffer](#6-how-to-increase-w5x00-txrx-buffer)
* [Examples](#examples)
  * [A. For Generic boards and shields](#a-for-generic-boards-and-shields)
    * [ 1. afraid-org_Client](examples/Generic/afraid-org_Client) 
    * [ 2. all-inkl_Client](examples/Generic/all-inkl_Client)
    * [ 3. DuckDNS_Client](examples/Generic/DuckDNS_Client)
    * [ 4. DynDNS_Client](examples/Generic/DynDNS_Client)
    * [ 5. dyndns-it_Client](examples/Generic/dyndns-it_Client)
    * [ 6. Dynu_Client](examples/Generic/Dynu_Client)
    * [ 7. enom_Client](examples/Generic/enom_Client)
    * [ 8. freemyip_Client](examples/Generic/freemyip_Client)
    * [ 9. No-ip_Client](examples/Generic/No-ip_Client)
    * [10. selfhost-de_Client](examples/Generic/selfhost-de_Client)
    * [11. strato_Client](examples/Generic/strato_Client)
    * [12. OVH_Client](examples/Generic/OVH_Client)
  * [B. For Ethernet shields](#b-for-ethernet-shields)
    * [ 1. nRF52_Ethernet_DuckDNS_Client](examples/Ethernet/nRF52/nRF52_Ethernet_DuckDNS_Client)
    * [ 2. SAMD_Ethernet_DuckDNS_Client](examples/Ethernet/SAMD/SAMD_Ethernet_DuckDNS_Client)
    * [ 3. SAM-DUE_Ethernet_DuckDNS_Client](examples/Ethernet/SAM-DUE/SAM-DUE_Ethernet_DuckDNS_Client)
    * [ 4. STM32_Ethernet_DuckDNS_Client](examples/Ethernet/STM32/STM32_Ethernet_DuckDNS_Client)
    * [ 5. AVR_Ethernet_DuckDNS_Client](examples/Ethernet/AVR/AVR_Ethernet_DuckDNS_Client)
    * [ 6. megaAVR_Ethernet_DuckDNS_Client](examples/Ethernet/megaAVR/megaAVR_Ethernet_DuckDNS_Client)
    * [ 7. Teensy_Ethernet_DuckDNS_Client](examples/Ethernet/Teensy/Teensy_Ethernet_DuckDNS_Client)
    * [ 8. RP2040_Ethernet_DuckDNS_Client](examples/Ethernet/RP2040/RP2040_Ethernet_DuckDNS_Client)
    * [ 9. WT32_ETH01_DuckDNS_Client](examples/Ethernet/WT32_ETH01/WT32_ETH01_DuckDNS_Client)
    * [10. Portenta_H7_Ethernet_DuckDNS_Client](examples/Ethernet/Portenta_H7/Portenta_H7_Ethernet_DuckDNS_Client). **New** 
  * [C. For WiFiNINA shields](#c-for-wifinina-shields)
    * [ 1. nRF52_WiFiNINA_DuckDNS_Client](examples/WiFiNINA/nRF52/nRF52_WiFiNINA_DuckDNS_Client)
    * [ 2. SAMD_WiFiNINA_DuckDNS_Client](examples/WiFiNINA/SAMD/SAMD_WiFiNINA_DuckDNS_Client)
    * [ 3. megaAVR_WiFiNINA_DuckDNS_Client](examples/WiFiNINA/megaAVR/megaAVR_WiFiNINA_DuckDNS_Client)
  * [D. For ESP8266-AT/ESP32-AT shields](#d-for-esp8266-atesp32-at-shields)
    * [ 1. nRF52_ESP_AT_DuckDNS_Client](examples/ESP_AT_WiFi/nRF52/nRF52_ESP_AT_DuckDNS_Client)
    * [ 2. SAMD_ESP_AT_DuckDNS_Client](examples/ESP_AT_WiFi/SAMD/SAMD_ESP_AT_DuckDNS_Client)
    * [ 3. SAM-DUE_ESP_AT_DuckDNS_Client](examples/ESP_AT_WiFi/SAM-DUE/SAM-DUE_ESP_AT_DuckDNS_Client)
    * [ 4. STM32_ESP_AT_DuckDNS_Client](examples/ESP_AT_WiFi/STM32/STM32_ESP_AT_DuckDNS_Client)
    * [ 5. AVR_ESP_AT_DuckDNS_Client](examples/ESP_AT_WiFi/AVR/AVR_ESP_AT_DuckDNS_Client).
    * [ 6. megaAVR_ESP_AT_DuckDNS_Client](examples/ESP_AT_WiFi/megaAVR/megaAVR_ESP_AT_DuckDNS_Client)
    * [ 7. Teensy_ESP_AT_DuckDNS_Client](examples/ESP_AT_WiFi/Teensy/Teensy_ESP_AT_DuckDNS_Client)
    * [ 8. RP2040_ESP_AT_DuckDNS_Client](examples/ESP_AT_WiFi/RP2040/RP2040_ESP_AT_DuckDNS_Client)
  * [E. For Portenta_H7 WiFi shields](#e-For-Portenta_H7-WiFi-shields)
    * [ 1. Portenta_H7_DuckDNS_Client](examples/WiFi/Portenta_H7/Portenta_H7_DuckDNS_Client). **New**
* [Example DuckDNS_Client](#example-duckdns_client)
  * [ 1. File DuckDNS_Client.ino](#1-file-duckdns_clientino) 
  * [ 2. File defines.h](#2-file-definesh)
  * [ 3. File wifi_credentials.h](#3-file-wifi_credentialsh)
* [Debug Terminal Output Samples](#debug-terminal-output-samples)
  *  [1. No-ip_Client on STM32F7 Nucleo-144 NUCLEO_F767ZI with LAN8742A using STM32Ethernet Library](#1-no-ip_client-on-stm32f7-nucleo-144-nucleo_f767zi-with-lan8742a-using-stm32ethernet-library)
  * [ 2. DuckDNS_Client on STM32F7 Nucleo-144 NUCLEO_F767ZI with LAN8742A using STM32Ethernet Library](#2-duckdns_client-on-stm32f7-nucleo-144-nucleo_f767zi-with-lan8742a-using-stm32ethernet-library)
  * [ 3. DuckDNS_Client on Adafruit NRF52840_FEATHER_EXPRESS with W5500 using Ethernet_Generic Library](#3-duckdns_client-on-adafruit-nrf52840_feather_express-with-w5500-using-Ethernet_Generic-library)
  * [ 4. No-ip_Client on Adafruit NRF52840_FEATHER_EXPRESS with W5500 using Ethernet_Generic Library](#4-no-ip_client-on-adafruit-nrf52840_feather_express-with-w5500-using-Ethernet_Generic-library)
  * [ 5. DuckDNS_Client on Adafruit NRF52840_FEATHER_EXPRESS with ENC28J60 using new EthernetENC Library](#5-duckdns_client-on-adafruit-nrf52840_feather_express-with-enc28j60-using-new-ethernetenc-library)
  * [ 6. DuckDNS_Client on Adafruit NRF52840_FEATHER_EXPRESS with ENC28J60 using UIPEthernet Library](#6-duckdns_client-on-adafruit-nrf52840_feather_express-with-enc28j60-using-uipethernet-library)
  * [ 7. DuckDNS_Client on Adafruit SAMD21 SAMD_NANO_33_IOT with WiFiNINA using WiFiNINA_Generic Library](#7-duckdns_client-on-adafruit-samd21-samd_nano_33_iot-with-wifinina-using-wifinina_generic-library)
  * [ 8. No-ip_Client on Adafruit SAMD21 SAMD_NANO_33_IOT with WiFiNINA using WiFiNINA_Generic Library](#8-no-ip_client-on-adafruit-samd21-samd_nano_33_iot-with-wifinina-using-wifinina_generic-library)
  * [ 9. DuckDNS_Client on Arduino SAM DUE with W5100 using Ethernet_Generic Library](#9-duckdns_client-on-arduino-sam-due-with-w5100-using-Ethernet_Generic-library)
  * [10. No-ip_Client on Arduino SAM DUE with W5100 using Ethernet_Generic Library](#10-no-ip_client-on-arduino-sam-due-with-w5100-using-Ethernet_Generic-library)
  * [11. DuckDNS_Client on SeeedStudio SEEED_XIAO_M0 with ESP8266-AT using ESP8266_AT_WebServer Library](#11-duckdns_client-on-seeedstudio-seeed_xiao_m0-with-esp8266-at-using-esp8266_at_webserver-library)
  * [12. No-ip_Client on SeeedStudio SEEED_XIAO_M0 with ESP8266-AT using ESP8266_AT_WebServer Library](#12-no-ip_client-on-seeedstudio-seeed_xiao_m0-with-esp8266-at-using-esp8266_at_webserver-library)
  * [13. AVR_Ethernet_DuckDNS_Client on AVR_MEGA2560 with W5x00 using Ethernet_Generic Library](#13-avr_ethernet_duckdns_client-on-avr_mega2560-with-w5x00-using-Ethernet_Generic-library)

changelog.md
# DDNS_Generic Library

[![arduino-library-badge](https://www.ardu-badge.com/badge/DDNS_Generic.svg?)](https://www.ardu-badge.com/DDNS_Generic)
[![GitHub release](https://img.shields.io/github/release/khoih-prog/DDNS_Generic.svg)](https://github.com/khoih-prog/DDNS_Generic/releases)
[![GitHub](https://img.shields.io/github/license/mashape/apistatus.svg)](https://github.com/khoih-prog/DDNS_Generic/blob/master/LICENSE)
[![contributions welcome](https://img.shields.io/badge/contributions-welcome-brightgreen.svg?style=flat)](#Contributing)
[![GitHub issues](https://img.shields.io/github/issues/khoih-prog/DDNS_Generic.svg)](http://github.com/khoih-prog/DDNS_Generic/issues)

---
---

## Table of Contents

* [Changelog](#changelog)
  * [Releases v1.7.0](#releases-v170)
  * [Releases v1.6.1](#releases-v161)
  * [Releases v1.6.0](#releases-v160)
  * [Releases v1.5.1](#releases-v151)
  * [Releases v1.5.0](#releases-v150)
  * [Releases v1.4.0](#releases-v140)
  * [Releases v1.3.0](#releases-v130)
  * [Releases v1.2.0](#releases-v120)
  * [Releases v1.1.0](#releases-v110)
  * [Releases v1.0.1](#releases-v101)
  * [Releases v1.0.0](#releases-v100)

---
---

## Changelog

### Releases v1.7.0

1. Use new [Ethernet_Generic library](https://github.com/khoih-prog/Ethernet_Generic) as default for W5x00.
2. Add support to `RP2040` using [`Earle Philhower's arduino-pico` core](https://github.com/earlephilhower/arduino-pico) with either `SPI0` or `SPI1`
3. Add support to WIZNet W5100S, such as  [**WIZnet Ethernet HAT**](https://docs.wiznet.io/Product/Open-Source-Hardware/wiznet_ethernet_hat) and [**W5100S-EVB-Pico**](https://docs.wiznet.io/Product/iEthernet/W5100S/w5100s-evb-pico)
4. Update `Packages' Patches`


### Releases v1.6.1

1. Auto detect ESP32 core version. Fix bug in examples for WT32_ETH01

### Releases v1.6.0

1. Replace deprecated `ipv4bot.whatismyipaddress.com` with `ifconfig.me`

### Releases v1.5.1

1. Update `platform.ini` and `library.json` to use original `khoih-prog` instead of `khoih.prog` after PIO fix
2. Update `Packages' Patches`

### Releases v1.5.0

1. Add support to **Portenta_H7**, using either WiFi or Vision-shield Ethernet
2. Add examples with new features

### Releases v1.4.0

1. Add support to **WT32_ETH01 (ESP32 + LAN8720)** boards
2. Add examples with new features

### Releases v1.3.0

1. Add support to RP2040-based boards, such as **RASPBERRY_PI_PICO, ADAFRUIT_FEATHER_RP2040 and GENERIC_RP2040**, using [Earle Philhower's arduino-pico core](https://github.com/earlephilhower/arduino-pico) and ESP8266/ESP32-AT WiFi
1. Add support to RP2040-based boards, such as **RASPBERRY_PI_PICO, ADAFRUIT_FEATHER_RP2040 and GENERIC_RP2040**, using [**Arduino-mbed RP2040** core](https://github.com/arduino/ArduinoCore-mbed) and Ethernet W5x00, ENC28J60.
3. Update examples with new features

### Releases v1.2.0

1. Add support to **Teensy boards** such as **Teensy LC, 3.0, 3.1, 3.2, 3.5, 3.6, 4.0, 4.1** using ESP-AT, WiFiNINA, W5x00 / ENC28J60 Ethernet
2. Add support to **Teensy 4.1 boards** using NativeEthernet


### Releases v1.1.0

1. Add support to **Arduino AVR ATmega2560** such as **Mega, Mega2560, Mega ADK, etc.**
2. Add support to **Arduino ATmega4809** such as **UNO WiFi Rev2, AVR_NANO_EVERY, etc.**
3. Remove dependency on <functional>.
4. Add OVH.com support.
5. Update and add examples for AVR and megaAVR
6. Add Version string
7. Add Table-of-Contents


### Releases v1.0.1

1. Add HTTP_ResponseParser class to fix issue with **nRF52 and STM32F/L/H/G/WB/MP1 using ESP8266/ESP32-AT**.
2. Add support to **Ethernet2, Ethernet3 and EthernetLarge libraries for STM32 using W5x00**.
3. Update Platform.ini to support PlatformIO 5.x owner-based dependency declaration.
4. Add more debug features.

#### Releases v1.0.0

1. Initial coding for Generic boards using many different WiFi/Ethernet modules/shields.
2. Add more examples




src/DDNS_Generic_Impl.h
/****************************************************************************************************************************
  DDNS_Generic_Impl.h
   
  For all Generic boards such as ESP8266, ESP32, SAM DUE, SAMD21/SAMD51, nRF52, STM32F/L/H/G/WB/MP1, AVR, megaAVR, Teensy
  with WiFiNINA, ESP8266/ESP32 WiFi, ESP8266-AT, W5x00, ENC28J60, built-in Ethernet LAN8742A, WT32_ETH01

  DDNS_Generic is a library to update DDNS IP address for DDNS services such as 
  duckdns, noip, dyndns, dynu, enom, all-inkl, selfhost.de, dyndns.it, strato, freemyip, afraid.org, ovh.com

  Based on and modified from 
  1) EasyDDNS            (https://github.com/ayushsharma82/EasyDDNS)
  2) ArduinoHttpClient   (https://github.com/arduino-libraries/ArduinoHttpClient)

  Built by Khoi Hoang https://github.com/khoih-prog/DDNS_Generic

  Licensed under MIT license
  
  Version: 1.7.0

  Version Modified By   Date      Comments
  ------- -----------  ---------- -----------
  1.0.0   K Hoang      11/09/2020 Initial coding for Generic boards using many WiFi/Ethernet modules/shields.
  1.0.1   K Hoang      28/09/2020 Fix issue with nRF52 and STM32F/L/H/G/WB/MP1 using ESP8266/ESP32-AT
  1.1.0   K Hoang      03/04/2021 Add OVH.com support. Remove dependency on <functional>. Add support to AVR Mega and megaAVR.
  1.2.0   K Hoang      04/04/2021 Add support to Teensy LC, 3.x, 4.0 and 4.1 using Ethernet, NativeEthernet, WiFi or ESP-AT
  1.3.0   K Hoang      15/05/2021 Add support to RP2040 using Ethernet or ESP-AT
  1.4.0   K Hoang      16/07/2021 Add support to WT32_ETH01 (ESP32 + LAN8720)
  1.5.0   K Hoang      16/09/2021 Add support to Portenta_H7, using either WiFi or Vision-shield Ethernet
  1.5.1   K Hoang      10/10/2021 Update `platform.ini` and `library.json`
  1.6.0   K Hoang      16/11/2021 Replace deprecated `ipv4bot.whatismyipaddress.com` with `ifconfig.me`
  1.6.1   K Hoang      01/12/2021 Auto detect ESP32 core version. Fix bug in examples for WT32_ETH01
  1.7.0   K Hoang      16/11/2021 Use new Ethernet_Generic library. Add support to RP2040 using arduino-pico code
 *****************************************************************************************************************************/

#ifndef DDNS_Generic_Impl_H
#define DDNS_Generic_Impl_H

void DDNSGenericClass::service(String ddns_service)
{
  ddns_choice = ddns_service;
}

void DDNSGenericClass::client(String ddns_domain, String ddns_username, String ddns_password)
{
  ddns_d = ddns_domain;
  ddns_u = ddns_username;
  ddns_p = ddns_password;
}

#if ( (WIFI_USE_STM32 || WIFI_USE_NRF528XX || WIFI_USE_AVR || WIFI_USE_MEGA_AVR || WIFI_USE_TEENSY || WIFI_USE_RP2040) && DDNS_USING_WIFI_AT )
// this method makes a HTTP connection to the server
String DDNSGenericClass::publicIPRequest(Client& client)
{
  // close any connection before send a new request
  // this will free the socket on the WiFi shield
  client.stop();

  // if there's a successful connection
  if (client.connect(server.c_str(), 80))
  {
    DDNS_LOGINFO1(F("Connecting to "), server);

    client.println("GET / HTTP/1.1");
    client.println("Host: " + server);
    client.println("Connection: close");
    client.println();
  }
  else
  {
    // if you couldn't make a connection
    DDNS_LOGERROR(F("Connection failed"));
  }
  
  String response = "";
  String currentIP;
  
  while ( response.length() == 0 )
  {
    response = "";
    
    // if there are incoming bytes available
    // from the server, read them and print them
    while (client.available())
    {
      char c = client.read();
      response += c;
    }

    DDNS_LOGDEBUG1(F("response = "), response);
    
    if (response.length() > 0)
    {
      HTTP_ResponseParser http = HTTP_ResponseParser();

      int httpCode = http.responseStatusCode(response);

      if (httpCode > 0)
      {
        if (httpCode == 200 /*HTTP_CODE_OK*/)
        {
          currentIP = http.responseBody(response);

          DDNS_LOGDEBUG1(F("Current Public IP = "), currentIP);
        }
      }
    }
  }
  
  client.stop();
  
  return currentIP;
}
#endif

void DDNSGenericClass::update(unsigned long ddns_update_interval, bool use_local_ip)
{
  interval = ddns_update_interval;

  unsigned long currentMillis = millis(); // Calculate Elapsed Time & Trigger

  if (( previousMillis == 0 ) || (currentMillis - previousMillis >= interval))
  {
#if (ESP8266 || ESP32 || DDNS_USING_WIFI)   
   #if DDNS_USING_WIFI_AT
      ESP8266_AT_Client client;
   #else   
      WiFiClient client;
   #endif   
#elif DDNS_USING_ETHERNET
      EthernetClient client;
#else
      WiFiClient client;   
#endif  

    previousMillis = currentMillis;

    if (use_local_ip)
    {
    
#if (ESP8266 || ESP32 || DDNS_USING_WIFI)    
      IPAddress ipAddress = WiFi.localIP();
#elif DDNS_USING_ETHERNET
      IPAddress ipAddress = Ethernet.localIP();
#else
      IPAddress ipAddress = WiFi.localIP();      
#endif

      new_ip = String(ipAddress[0]) + String(".") +
               String(ipAddress[1]) + String(".") +
               String(ipAddress[2]) + String(".") +
               String(ipAddress[3]);
               
      DDNS_LOGINFO1(F("Current Local IP ="), new_ip);         
    } 
    
      
#if (ESP8266 || ESP32)
    else 
    {
      // ######## GET PUBLIC IP ######## //
  
      HTTPClient http;

      http.begin(client, "http://ifconfig.me/ip");
      
      DDNS_LOGDEBUG(F("Calling http.begin"));
      
      // @return -1 if no info or > 0 when Content-Length is set by server
      int httpCode = http.GET();
      
      DDNS_LOGDEBUG1(F("httpCode ="), httpCode);
      
      if (httpCode > 0) 
      {
        if (httpCode == HTTP_CODE_OK) 
        {
          new_ip = http.getString();
          
          DDNS_LOGINFO1(F("Current Public IP ="), new_ip);
        }
      } 
      else 
      {
        http.end();
        
        return;
      }
      
      http.end();
    }
    
#else

    else 
    {
#if ( (WIFI_USE_STM32 || WIFI_USE_NRF528XX || WIFI_USE_AVR || WIFI_USE_MEGA_AVR || WIFI_USE_TEENSY || WIFI_USE_RP2040) && DDNS_USING_WIFI_AT )
      // To fix issue on nRF52 and STM32 using ESP8266/ESP32-AT
      // ######## GET PUBLIC IP ######## //
      DDNS_LOGDEBUG(F("Calling publicIPRequest"));
       
      new_ip = publicIPRequest(client);
      
      if (new_ip.length() > 0)
        DDNS_LOGERROR1(F("Current Public IP ="), new_ip);
#else

      // ######## GET PUBLIC IP ######## //
      HttpClient http(client, "ifconfig.me");

      DDNS_LOGERROR("Access ifconfig.me");
      
      // @return HTTP_CODE_OK (200) if OK and when responseBody is responded by server
      http.get("/ip");
      
      int httpCode = http.responseStatusCode();
      
      DDNS_LOGERROR1("httpCode =", httpCode);  
      
      if (httpCode > 0)
      {
        if (httpCode == 200 /*HTTP_CODE_OK*/) 
        {     
          String response = http.responseBody();
          new_ip = response;
          
          DDNS_LOGERROR1("Current Public IP =", new_ip);
          DDNS_LOGERROR1("response =", response);      
        }
      } 
      else 
      {
        DDNS_LOGERROR("http.stop");
        
        http.stop();
        
        return;
      }
      
      http.stop();      
      
      
#endif
      
    } 
        
#endif      
        
    // ######## GENERATE UPDATE URL ######## //
    
    String access_url;
    String access_path;
    
#if !(ESP8266 || ESP32)
    bool needAuth;
    
    if ( (ddns_choice == "noip") || (ddns_choice == "dyndns") || (ddns_choice == "all-inkl") || (ddns_choice == "selfhost.de") || 
         (ddns_choice == "dyndns.it") || (ddns_choice == "strato") || (ddns_choice == "ovh") || (ddns_choice == "dyndns.it") )
      needAuth = true;
    else
      needAuth = false;
#endif
        
    if (ddns_choice == "duckdns") 
    {
      access_url  = "www.duckdns.org";
      access_path = "/update?domains=" + ddns_d + "&token=" + ddns_u + "&ip=" + new_ip + "";
      
      update_url = "http://www.duckdns.org/update?domains=" + ddns_d + "&token=" + ddns_u + "&ip=" + new_ip + "";
    } 
    else if (ddns_choice == "noip") 
    {     
      access_url  = "dynupdate.no-ip.com";
      access_path = "/nic/update?hostname=" + ddns_d + "&myip=" + new_ip + "";
      
      update_url = "http://" + ddns_u + ":" + ddns_p + "@dynupdate.no-ip.com/nic/update?hostname=" + ddns_d + "&myip=" + new_ip + "";
    } 
    else if (ddns_choice == "dyndns") 
    {
      access_url  = "members.dyndns.org";
      access_path = "/v3/update?hostname=" + ddns_d + "&myip=" + new_ip + "";
      
      update_url = "http://" + ddns_u + ":" + ddns_p + "@members.dyndns.org/v3/update?hostname=" + ddns_d + "&myip=" + new_ip + "";
    } 
    else if (ddns_choice == "dynu") 
    {
      access_url  = "api.dynu.com";
      access_path = "/nic/update?hostname=" + ddns_d + "&myip=" + new_ip + "&username=" + ddns_u + "&password=" + ddns_p + "";
      
      update_url = "http://api.dynu.com/nic/update?hostname=" + ddns_d + "&myip=" + new_ip + "&username=" + ddns_u + "&password=" + ddns_p + "";
    } 
    else if (ddns_choice == "enom") 
    {
      access_url  = "dynamic.name-services.com";
      access_path = "/interface.asp?command=SetDnsHost&HostName=" + ddns_d + "&Zone=" + ddns_u + "&DomainPassword=" + ddns_p + "&Address=" + new_ip + "";
      
      update_url = "http://dynamic.name-services.com/interface.asp?command=SetDnsHost&HostName=" + ddns_d + "&Zone=" + ddns_u + "&DomainPassword=" + ddns_p + "&Address=" + new_ip + "";
    } 
    else if (ddns_choice == "all-inkl") 
    {      
      access_url  = "dyndns.kasserver.com";
      access_path = "/?myip=" + new_ip;
      
      update_url = "http://" + ddns_u + ":" + ddns_p + "@dyndns.kasserver.com/?myip=" + new_ip;
    } 
    else if (ddns_choice == "selfhost.de") 
    {     
      access_url  = "carol.selfhost.de";
      access_path = "/nic/update?";
      
      update_url = "http://" + ddns_u + ":" + ddns_p + "@carol.selfhost.de/nic/update?";
    } 
    else if (ddns_choice == "dyndns.it") 
    {     
      access_url  = "update.dyndns.it";
      access_path = "/nic/update?hostname=" + ddns_d;
      
      update_url = "http://" + ddns_u + ":" + ddns_p + "@update.dyndns.it/nic/update?hostname=" + ddns_d;
    } 
    else if (ddns_choice == "strato") 
    {      
      access_url  = "dyndns.strato.com";
      access_path = "/nic/update?hostname=" + ddns_d + "&myip=" + new_ip + "";
      
      update_url = "http://" + ddns_u + ":" + ddns_p + "@dyndns.strato.com/nic/update?hostname=" + ddns_d + "&myip=" + new_ip + "";
    } 
    else if (ddns_choice == "freemyip") 
    {
      access_url  = "freemyip.com";
      access_path = "/update?domain=" + ddns_d + "&token=" + ddns_u + "&myip=" + new_ip + "";
      
      update_url = "http://freemyip.com/update?domain=" + ddns_d + "&token=" + ddns_u + "&myip=" + new_ip + "";
    } 
    else if (ddns_choice == "afraid.org") 
    {
      access_url  = "sync.afraid.org";
      access_path = "/u/" + ddns_u + "/";
      
      update_url = "http://sync.afraid.org/u/" + ddns_u + "/";
    }
    else if (ddns_choice == "ovh") 
    {      
      access_url  = "www.ovh.com";
      access_path = "/nic/update?system=dyndns&hostname=" + ddns_d + "&myip=" + new_ip + "";
      
      update_url = "http://" + ddns_u + ":" + ddns_p + "@www.ovh.com/nic/update?system=dyndns&hostname=" + ddns_d + "&myip=" + new_ip + "";
    }
    else 
    {
      DDNS_LOGERROR(F("## INPUT CORRECT DDNS SERVICE NAME ##"));
      return;
    }

    // ######## CHECK & UPDATE ######### //
    if ( (new_ip.length() > 0) && (old_ip != new_ip) )
    {
    
#if (ESP8266 || ESP32)
      HTTPClient http;
      
      DDNS_LOGWARN1(F("Sending HTTP_GET to"), ddns_choice);
      DDNS_LOGWARN1(F("HTTP_GET ="),  update_url);
      
      http.begin(client, update_url);
      
      // @return HTTP_CODE_OK (200) if OK and when responseBody is responded by server
      int httpCode = http.GET();    
      
      if (httpCode == HTTP_CODE_OK) 
      {
        // Send a callback notification
        if (_ddnsUpdateFunc != nullptr) 
        {
          _ddnsUpdateFunc(old_ip.c_str(), new_ip.c_str());
        }
        
        // Replace Old IP with new one to detect further changes.
        old_ip = new_ip;
        
        DDNS_LOGERROR1(F("Updated IP ="), old_ip);
      }
      else if (httpCode == -1)
      {
        DDNS_LOGERROR(F("Public IP not changed. Don't need to update"));
      }
      
      http.end();
#else
     
      HttpClient http(client, access_url);
      
      if (needAuth)
      {
        http.beginRequest();
        http.get(access_path);
        http.sendBasicAuth(ddns_u, ddns_p); // send the username and password for authentication
        http.endRequest();
      }
      else
      { 
        // @return -1 if no info or > 0 when Content-Length is set by server
        http.get(access_path);
      }

      DDNS_LOGWARN1(F("Sending HTTP_GET to"), ddns_choice);
      DDNS_LOGWARN1(F("HTTP_GET ="),  update_url);
           
      // @return 200 when Content-Length is set by server
      int httpCode = http.responseStatusCode();
      
      DDNS_LOGINFO1(F("httpCode ="), httpCode);  
      
      if (httpCode == 200 /*HTTP_CODE_OK*/) 
      {
        // Send a callback notification
        if (_ddnsUpdateFunc != nullptr) 
        {
          _ddnsUpdateFunc(old_ip.c_str(), new_ip.c_str());
        }
      
        // Replace Old IP with new one to detect further changes.
        old_ip = new_ip;
        
        DDNS_LOGERROR1(F("Updated IP ="), old_ip); 
      }
      else if (httpCode == -1)
      {
        DDNS_LOGERROR(F("Public IP not changed. Don't need to update"));
      }
      
      http.stop();
#endif      
    }
  }
}

DDNSGenericClass DDNSGeneric;

#endif    // DDNS_Generic_Impl_H
